### ISSUE TITLE
What guarantees are provided by repr(packed)?

### ISSUE THREAD

**Issue by joshlf** [rust-lang/unsafe-code-guidelines#318]
[Per the nomicon](https://doc.rust-lang.org/nomicon/other-reprs.html#reprpacked), `repr(packed)` guarantees that all inter-field padding is removed. However, [the reference](https://doc.rust-lang.org/stable/reference/type-layout.html) merely says that "`packed` **may** also alter the padding between fields" (emphasis added).

What guarantees does `repr(packed)` provide? In particular:
- Does `repr(packed)` guarantee that all inter-field padding is removed?
- Does `repr(packed)` guarantee that a type's layout is fixed in any way _other_ than removing padding (e.g., field ordering)?

**Comment by RalfJung** [comment]
> Does repr(packed) guarantee that all inter-field padding is removed?

I would strongly think so, that is the entire point after all. Things become a bit more tricky to describe with `packed(N)` though. Still I think this can be just fixed via a PR against the reference.

As for field ordering, I don't know if it has any effect.

**Comment by Diggsey** [comment]
`#[repr(packed)]` means the struct won't add any padding as part of its own layout, but there might still effectively be padding between fields, eg. if the type of a field has size=4, alignment=8, then its my understanding that there would still be 4 bytes of padding because that padding belongs to the field itself, and is not added by the struct.

**Comment by RalfJung** [comment]
Yes, the padding contained inside field types (including trailing padding to make the size a multiple of the alignment) is unaffected.

**Comment by joshlf** [comment]
Submitted: https://github.com/rust-lang/reference/pull/1163
